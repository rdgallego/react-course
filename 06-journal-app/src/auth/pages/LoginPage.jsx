import { Google } from "@mui/icons-material"
import { Alert, Button, Grid, Link, TextField, Typography } from "@mui/material"
import { Link as RouterLink } from "react-router-dom";
import { AuthLayout } from "../layout/AuthLayout";
import { useForm } from "../../hooks";
import { startGoogleSignIn, startLoginWithEmailPassword } from "../../store/auth/thunks";
import { useDispatch, useSelector } from "react-redux";
import { useMemo } from "react";

const formData = {
  email: '',
  password: ''
}
export const LoginPage = () => {

  const {status, errorMessage} = useSelector(state => state.auth)
  const dispatch = useDispatch()

  const isAuthenticating = useMemo(() => status === 'checking', [status])

  const {onInputChange, email, password} = useForm(formData)

  const onSubmit = (event) => {
    event.preventDefault();

    dispatch(startLoginWithEmailPassword({email, password}));
  }

  const onGoogleSignIn = () => {
    dispatch(startGoogleSignIn());
  }
  return (
    <AuthLayout title="Login">
      <form data-testid="form" onSubmit={onSubmit} className="animate__animated animate__fadeIn animate__faster">
          <Grid container spacing={2} sx={{mb: 2, mt: 1}}>
            <Grid item xs={12}>
              <TextField label="Correo" type="email" placeholder="correo@google.com" name="email" value={email} onChange={onInputChange} fullWidth></TextField>
            </Grid>
            <Grid item xs={12}>
              <TextField inputProps={{'data-testid': 'Contraseña'}} label="Contraseña" type="password" placeholder="Contraseña" name="password" value={password} onChange={onInputChange} fullWidth></TextField>
            </Grid>
          </Grid>

          <Grid container spacing={2} sx={{mb: 2}}>
          <Grid item xs={12} display={!!errorMessage ? '' : 'none'}>
            <Alert severity="error">{errorMessage}</Alert>
          </Grid>
            <Grid item xs={12} sm={6}>
              <Button type="submit" variant="contained" fullWidth disabled={isAuthenticating}>
                Login
              </Button>
            </Grid>
            <Grid item xs={12} sm={6}>
              <Button aria-label="google-btn" onClick={onGoogleSignIn} variant="contained" fullWidth disabled={isAuthenticating}>
                <Google  />
                <Typography sx={{ml: 1}}>Google</Typography>
              </Button>
            </Grid>

            <Grid container direction="row" justifyContent="end">
              <Link component={RouterLink} color="inherit" to="/auth/register">
                Crear una cuenta
              </Link>
            </Grid>
          </Grid>
        </form>
    </AuthLayout>
  )
}
