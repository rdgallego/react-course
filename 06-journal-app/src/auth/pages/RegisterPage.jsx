import { Alert, Button, Grid, Link, TextField } from "@mui/material"
import { Link as RouterLink } from "react-router-dom";
import { AuthLayout } from "../layout/AuthLayout";
import { useForm } from "../../hooks";
import { useMemo, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { startCreatingUserWithEmailPassword } from "../../store/auth/thunks";
const formData = {
  email: '',
  password: '',
  displayName: ''
}

const formValidations = {
  email: [(value) => value.includes('@'), 'El correo debe de tener un @'],
  password: [(value) => value.length >= 6, 'La contraseña debe de tener mas de 6 letras'],
  displayName: [(value) => value.length >= 1, 'El nombre es obligatorio'],
}
export const RegisterPage = () => {

  const {status, errorMessage }= useSelector(state => state.auth)
  const isCheckingAuthentication = useMemo( () => status === 'checking', [status])


  const {onInputChange, email, password, displayName, emailValid, passwordValid, displayNameValid, isFormValid, formState} = useForm(formData, formValidations)

  const [formSubmitted, setFormSubmitted] = useState(false)

  const dispatch = useDispatch();

  const onSubmit = (event) => {
    event.preventDefault();
    setFormSubmitted(true);

    if(!isFormValid) return;

    dispatch(startCreatingUserWithEmailPassword(formState))
  }

  return (
    <AuthLayout title="Register">
    <form onSubmit={onSubmit} className="animate__animated animate__fadeIn animate__faster">
        <Grid container spacing={2} sx={{mb: 2, mt: 1}}>
          <Grid item xs={12}>
            <TextField
            label="Nombre"
            type="text"
            placeholder="Nombre"
            fullWidth
            name="displayName"
            value={displayName}
            onChange={onInputChange}
            error={!!displayNameValid && formSubmitted}
            helperText={displayNameValid}></TextField>
          </Grid>
          <Grid item xs={12}>
            <TextField
            label="Correo"
            type="email"
            placeholder="correo@google.com"
            fullWidth
            name="email"
            value={email}
            onChange={onInputChange}
            error={!!emailValid && formSubmitted}
            helperText={emailValid}></TextField>
          </Grid>
          <Grid item xs={12}>
            <TextField
            label="Contraseña"
            type="password"
            placeholder="Contraseña"
            fullWidth
            name="password"
            value={password}
            onChange={onInputChange}
            error={!!passwordValid && formSubmitted}
            helperText={passwordValid}></TextField>
          </Grid>
        </Grid>

        <Grid container spacing={2} sx={{mb: 2}}>
          <Grid item xs={12} display={!!errorMessage ? '' : 'none'}>
            <Alert severity="error">{errorMessage}</Alert>
          </Grid>
          <Grid item xs={12} sm={12}>
            <Button variant="contained" fullWidth type="submit" disabled={isCheckingAuthentication}>
              Register
            </Button>
          </Grid>

          <Grid container direction="row" justifyContent="end">
            <Link component={RouterLink} color="inherit" to="/auth/login">
              Ya tengo cuenta
            </Link>
          </Grid>
        </Grid>
      </form>
  </AuthLayout>
  )
}
