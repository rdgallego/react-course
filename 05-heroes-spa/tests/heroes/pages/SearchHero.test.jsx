import { fireEvent, render, screen } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"
import { SearchHero } from "../../../src/heroes/pages/SearchHero"

/* eslint-disable no-undef */

const mockedUseNavigate = jest.fn()
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate
}))
describe('pruebas en search hero', () => { 
    test('should show default values', () => { 
        const {container} = render(
            <MemoryRouter>
                <SearchHero/>
            </MemoryRouter>
        )

        expect(container).toMatchSnapshot();
     })

     test('should show batman', () => { 
        render(
            <MemoryRouter initialEntries={['/search?q=Batman']}>
                <SearchHero/>
            </MemoryRouter>
        )
        const input = screen.getByRole('textbox')
        expect(input.value).toBe('Batman');

        const img = screen.getByRole('img')
        expect(img.src).toContain('/assets/heroes/dc-batman.jpg');
      })

      test('should show error if cant find hero', () => { 
            render(
                <MemoryRouter initialEntries={['/search?q=cantfind']}>
                    <SearchHero/>
                </MemoryRouter>
            )
            const input = screen.getByRole('textbox')
            expect(input.value).toBe('cantfind');

            expect(screen.findByText('No hero with cantfind')).toBeTruthy();
       })

      test('should call navigate to new page ', () => { 
            render(
                <MemoryRouter initialEntries={['/search']}>
                    <SearchHero/>
                </MemoryRouter>
            )
            const input = screen.getByRole('textbox')
            fireEvent.input(input, { target: {value: 'cantfind'} });

            const form = screen.getByLabelText('form');
            fireEvent.submit(form);
            expect(mockedUseNavigate).toHaveBeenCalledWith('?q=cantfind');
       })
 })