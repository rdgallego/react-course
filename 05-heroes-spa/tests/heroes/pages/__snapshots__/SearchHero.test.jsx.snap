// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en search hero should show default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Seach
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Result
      </h4>
      <hr />
      <div
        class="alert alert-danger"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
